Fix cross eval issues

Artturin opened the pull request:
Split from https://github.com/NixOS/nixpkgs/pull/341067

Exposed by https://github.com/NixOS/nixpkgs/commit/bd32adb9e8452e96f7dbb5ee0bebd96bab0ea320.patch
https://github.com/NixOS/nixpkgs/commit/6761b8e2ff4861cd5f371ebb2753ec547078221d.patch

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Artturin committed:
python2: Fix alias usage on cross

Artturin committed:
saleae-logic: Use platforms instead of assert

Artturin committed:
haskellPackages.mkDerivation: Fix `cannot coerce null to a string`

https://www.github.com/NixOS/nixpkgs/issues/362272

Fixes

```
$ nix repl --file . --system aarch64-darwin
nix-repl> pkgsCross.aarch64-multiplatform.haskellPackages.ghc-lib.env
error:
       … while evaluating attribute 'LOCALE_ARCHIVE' of derivation 'ghc-shell-for-ghc-lib-aarch64-unknown-linux-gnu-9.6.6.20240701-aarch64-unknown-linux-gnu'
       error: cannot coerce null to a string: null
```

Artturin committed:
treewide: Change unnecessary aborts to throws

`throw`s are better because they don't abort evaluation when evaluating the whole set.

Fixes

`error: evaluation aborted with the following error message: 'Unsupported platform: arm64.'` on linux cross from darwin

Artturin committed:
tests-stdenv-gcc-stageCompare: Fail gracefullyier

Don't fail with a missing attribute error, so the error can be handled.

mweinelt reviewed (approved):
Python changes look fine. Could use python27 for one less indirection.

philiptaron commented on pkgs/development/interpreters/python/cpython/2.7/default.nix:
```diff
@@ -204,7 +204,7 @@ let
     "--disable-toolbox-glue"
   ] ++ lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform) [
-    "PYTHON_FOR_BUILD=${lib.getBin buildPackages.python}/bin/python"
+    "PYTHON_FOR_BUILD=${lib.getBin buildPackages.python2}/bin/python"
```
```suggestion
    "PYTHON_FOR_BUILD=${lib.getBin buildPackages.python27}/bin/python"
```

philiptaron replied:
Strangely, `buildPackages.python27` doesn't have `meta.mainProgram` set, so `lib.getExe buildPackages.python27` doesn't work.

philiptaron commented on pkgs/development/interpreters/python/cpython/2.7/default.nix:
```diff
@@ -239,7 +239,7 @@ let
     [ autoreconfHook ]
     ++ lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform)
-      [ buildPackages.stdenv.cc buildPackages.python ];
+      [ buildPackages.stdenv.cc buildPackages.python2 ];
```
```suggestion
      [ buildPackages.stdenv.cc buildPackages.python27 ];
```

philiptaron reviewed (approved):
I like @mweinelt's suggestion about `python2` -> `python27`.

sternenseemann commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -802,7 +802,7 @@ stdenv.mkDerivation ({
           otherBuildInputsSystem;
         LANG = "en_US.UTF-8";
-        LOCALE_ARCHIVE = lib.optionalString (stdenv.hostPlatform.libc == "glibc") "${buildPackages.glibcLocales}/lib/locale/locale-archive";
+        LOCALE_ARCHIVE = lib.optionalString (stdenv.hostPlatform.libc == "glibc" && buildPackages.glibcLocales != null) "${buildPackages.glibcLocales}/lib/locale/locale-archive";
```
```suggestion
        LOCALE_ARCHIVE = lib.optionalString (stdenv.buildPlatform.libc == "glibc") "${buildPackages.glibcLocales}/lib/locale/locale-archive";
```

This would make the condition the same we use in the actual builder, probably old code that never got fixed. `LOCALE_ARCHIVE` only influences the runtime so it makes no sense to set it based on host in the build environment.

https://github.com/NixOS/nixpkgs/blob/e1bdae5eb8c5efd0da2ba8882c257b4a73137a58/pkgs/development/haskell-modules/generic-builder.nix#L855

Artturin committed:
frr: Fix cross `snmpSupport` to remove warning which interferes with cross eval

Release eval fails with

```
nix-env failed:
trace: evaluation warning: cannot enable SNMP support due to cross-compilation issues with net-snmp-config
```

philiptaron reviewed (approved):
Lgtm % @sternenseemann saying yes

Artturin committed:
reposilitePlugins: Fix eval

Artturin committed:
lib/tests/release.nix: Fix build if `impure.nix` returns a cross pkgs

In https://www.github.com/NixOS/nixpkgs/pull/341067 `ofborg-eval-lib-tests`
fails because it uses `pkgsHostTarget` of the cross pkgs

```diff
diff --git a/pkgs/top-level/impure.nix b/pkgs/top-level/impure.nix
index 4d847e280f4b..7f0099b254f0 100644
--- a/pkgs/top-level/impure.nix
+++ b/pkgs/top-level/impure.nix
@@ -86,4 +86,6 @@ assert args ? system -> !(args ? localSystem);

 import ./. (builtins.removeAttrs args [ "system" ] // {
   inherit config overlays localSystem;
-})
+} // (if ((localSystem.system or localSystem) == "x86_64-linux" || (localSystem.system or localSystem) == "aarch64-linux") then {
+  crossSystem = { config = "aarch64-unknown-linux-gnu"; };
+  } else {}) )
```

philiptaron reviewed (approved):
Latest changes look good to me.

Artturin committed:
nixpkgs-manual.{lib-docs,tests}: Fix cross build

sternenseemann commented on pkgs/by-name/fr/frr/package.nix:
```diff
@@ -81,9 +81,6 @@
 , ospfApi ? true
 }:
 
-lib.warnIf (!(stdenv.buildPlatform.canExecute stdenv.hostPlatform))
```
If explicitly enabling `snmpSupport`, does the build fail or is the feature silently disabled?

Artturin replied:
Fails without finding `net-snmp-config`, I now just fixed snmpSupport on cross  [`9f2266d` (#362546)](https://github.com/NixOS/nixpkgs/pull/362546/commits/9f2266dba97a2df361614788c8cd76738386e0f5)

sternenseemann replied:
That's even better! Just wanted to make sure no one's expectations are silently subverted…

Artturin commented on pkgs/development/compilers/swift/compiler/default.nix:
```diff
@@ -47,6 +47,8 @@ let
     else
       targetPlatform.darwinMinVersion;
 
+  targetDarwinPlatform = if (targetPlatform.darwinPlatform != null) then targetPlatform.darwinPlatform else "null";
```
I don't think so, but it was still being evaluated, see commit msg.

sternenseemann replied:
Maybe the `hostPlatform.isDarwin` check surrounding the second `targetDarwinPlatform` is also wrong?

Artturin replied:
Removed the other changes and only changed the `hostPlatform.isDarwin` to `targetPlatform.isDarwin`. There are other conditionals which are probably wrong but not going to touch them.

Artturin commented on pkgs/test/stdenv/gcc-stageCompare.nix:
```diff
@@ -14,6 +14,8 @@
 }:
 
 assert stdenv.cc.isGNU;
+assert lib.assertMsg (pkgs.gcc-unwrapped ? checksum) "No `checksum` output in `gcc-unwrapped` see conditional in `gcc/common/checksum.nix`";
```
You mean add a stub `checksum` attr to gcc? I think it's fine the way it is now, we don't stub conditional outputs in other packages.

sternenseemann replied:
I meant to check after the `let` on the actual `gcc-stageCompare` thing instead of the `gcc-unwrapped` that we don't actually used (but override).

Artturin replied:
Changed.

Artturin committed:
swift: Fix eval failure `error: expected a string but found null: null`

Dunno if this is correct, there's other `hostPlatform.isDarwin`
conditionals which may be wrong.

Leaving the other conditionals to swift maintainers or people who are
interested in swift cross-compilation

```
$ nix repl --file . --system aarch64-darwin
nix-repl> pkgsCross.aarch64-multiplatform.swift
«derivation /nix/store/kqbaqwdfc9g0yvkq4ffrlqspranprlmy-swift-wrapper-aarch64-unknown-linux-gnu-5.8.drv»

nix-repl> pkgsCross.aarch64-multiplatform.buildPackages.swift
error:
       error: expected a string but found null: null
       at /home/artturin/nixgits/my-nixpkgs/.worktree/2/pkgs/development/compilers/swift/compiler/default.nix:586:11:
          585|         #WATCHOS_SIMULATOR
          586|       }.${targetPlatform.darwinPlatform}}
             |           ^
          587|
```

sternenseemann commented:
`lib/tests/release.nix` is not pretty, but honestly I don't know what to do there either.

Artturin commented:
Added comment to release.nix and updated the patch in the commit msg
